### PR TITLE
Introduce registry of XML models

### PR DESCRIPTION
**Pull request by jonseymour**
This series introduces a registry of XML models.

The XML models contain metadata that is required to successfully map a stream of bytes onto a Block{N} or Block{N}Repeat structure and vice versa and to apply scaling factors as and when required.

An alternative to the current placement of the generated logic for the model registrations would be to run a separate generator that creates a models.go file in the smdx package.

### CODE REVIEW

**Comment by jonseymour** [comment]
Thinking about this some more...

It is probably uncool to load the entire Sunspec model space into every Go executable particularly if you are only intend to use a couple of models in a given context. One way to fix this would be to put each model registration into its own go package along with constants that name each point, that way the Go compiler can remove models that are not actually used at compile time.

Also, it isn't clear to me that defining Block{N} structures is really that useful in practice for the following reasons. While it *might* be reasonable to read an entire model into memory, it is almost never correct to write a whole model and it isn't clear how partial updates written directly into a Block{N} structure would be  intercepted and converted into Modbus write operations. Further the Go memory layout will not be byte-for-byte identical with the wire format of a block because of byte ordering issues and the way that strings are (currently) represented in the Block{N} structs. This means that there is no technical advantage to defining a contiguous block structure, at least on the Modbus side.

It seems to me that a better way for a programmer to interact with a model would be to specify, with symbolic identifiers, the set of points to be read, read the registers corresponding to those points into a buffer, and then use the point identifier with strongly typed accessor methods to read and write the points of interest to/from the buffer of interest.

So: 

        Block b;

        b.Prepare(model1.Mn, model1.DA)
        b.Refresh()  // executes a sequence of ReadHoldingRegisters calls to populate a buffer
        da := b.GetUint16(model1.DA) // unpacks the buffer

Similarly for write:

        b.PutUint16(model1.DA, 1) // write into a local buffer
        b.Flush() // flush the buffer into WriteHoldingRegisters calls

So, what I plan to do is this:

 - change the generator to generate one package under core/models for each model
 - in the model specific package, define constants for each point name
 - in the model specific package, register the XML model
 - ~~don't generate a Block{N} or Block{N}Repeat struct for each model~~
 - define an API for reading and writing block buffers from an underlying Modbus API

**Updated:** thinking about it some more, I guess the Block types might be useful place for an application programmer to store partial reads that return multiple points even if full reads and writes of a block aren't always practical in themselves. So I have kept the block types around in the reorganised generation.

**Comment by jonseymour** [comment]
I am wondering if the models might better be generated within packages named like "sunspec/model{N}" or "sunspec/models/model{N}" rather than "core/models/model{N}"?

File: generators/core.go [hunk 1]
```diff
@@ -10,81 +10,97 @@ import (
...
-	Name: "{{.Model.Name}}",
-	Length: {{.Model.Length}},
-	Blocks: []smdx.BlockElement{ {{range .Model.Blocks}}
+func init() {
+	smdx.RegisterModel(&smdx.ModelElement{
+	Id: {{$model.Id}},
+	Name: "{{$model.Name}}",
+	Length: {{$model.Length}},
+	Blocks: []smdx.BlockElement{ {{range $model.Blocks}}
 			smdx.BlockElement{ {{if gt (len .Name) 0 }}Name: "{{.Name}}",{{end}}
```
Comment (crabmusket) [inline generators/core.go]: Not that it's a huge deal, but the output with `BlockElement{Name: ""` is a bit odd

**Comment by crabmusket** [comment]
* I agree with all your points about the `Block{N}` structs. The original intent of those types was for use in application-side code (e.g. analysis of the data stream in a monitoring device) but that can be accomplished using your buffer model.
* The namespace is nested under `github.com/crabmusket/gosunspec/`, so I'm not a fan of repeating `sunspec`, but `core` isn't very descriptive so a change might be for the best.

This looks good! Is it mergeable? Given I'm so distant from the code, I'll basically merge anything that works for you. But I will try to at least give a cursory review of everything before I do!

File: generators/core.go (hunk 1 above)
Comment (jonseymour) [reply generators/core.go]: The construction above should explicitly prevent `BlockElement{Name: ""` while allowing `BlockElement{Name: "foobar"`. 

That said, I think it is quite possible I didn't understand the actual meaning of your comment.

**Comment by jonseymour** [comment]
Ok, so what I have done with these last 3 commits, is reorganise the namespace so that the models are now under `models/ `rather than `core/models/`.

To make this work, I have to put the SMDX modules into `spec/`, to remove the conflict that would otherwise arise.

I have also made `spec/` a submodule which has the advantage of allowing the particular spec used to generate a model tree to be tied exactly to the model tree it generates.

Subsequent commits will fix the compilation errors in`xml` which I hadn't noticed until now.

**Comment by jonseymour** [comment]
I decided to remove the core.String type and replace all uses of it with the golang string type based on ease of use considerations.

**Comment by crabmusket** [comment]
Seems fine. `core.String` was one of those things I wasn't sure about. I think there might have been a couple of other types that I didn't alias to core types as well.

I like the idea of making the spec a submodule, good thinking.

Let me know when this is mergeable :)

**Comment by jonseymour** [comment]
It is ready to merge in its current state.

Something that might be worth doing is moving the core/* types into the top level, then changing the package from core to "sunspec". That way everyone gets the sunspec prefix by default. If you'd like me to do that, I can do that next.

I am currently working on an interface model that should be general enough to allow working with models that are backed by XML documents or by a modbus interface, but I won't push that until these changes have been merged.

**Comment by crabmusket** [comment]
Prefix by default would probably be good :+1:

**Pull request merged**